xdsl/dialects/riscv.py

```diff
@@ -0,0 +1,765 @@
+class MULOp(Riscv1Rd2RsOperation):
+    name = "riscv.mul"
+
+
+@irdl_op_definition
+class MULHOp(Riscv1Rd2RsOperation):
+    name = "riscv.mulh"
+
+
+@irdl_op_definition
+class MULHSUOp(Riscv1Rd2RsOperation):
+    name = "riscv.mulhsu"
+
+
+@irdl_op_definition
+class MULHUOp(Riscv1Rd2RsOperation):
+    name = "riscv.mulhu"
+
+
+@irdl_op_definition
+class DIVOp(Riscv1Rd2RsOperation):
+    name = "riscv.div"
+
+
+@irdl_op_definition
+class DIVUOp(Riscv1Rd2RsOperation):
+    name = "riscv.divu"
+
+
+@irdl_op_definition
+class REMOp(Riscv1Rd2RsOperation):
+    name = "riscv.rem"
+
+
+@irdl_op_definition
+class REMUOp(Riscv1Rd2RsOperation):
+    name = "riscv.remu"
+
+
+# Pseudo Ops (Table 25.2)
+#
+# We do not include sext.w as we cannot parse this operation yet
+# and also do not include the floating point operations.
+
+
+@irdl_op_definition
+class NOPOp(RiscvNoParamsOperation):
+    name = "riscv.nop"
+
+
+@irdl_op_definition
+class LIOp(Riscv1Rd1ImmOperation):
+    name = "riscv.li"
+
+
+@irdl_op_definition
+class MVOp(Riscv1Rd1RsOperation):
+    name = "riscv.mv"
+
+
+@irdl_op_definition
+class NOTOp(Riscv1Rd1RsOperation):
+    name = "riscv.not"
+
+
+@irdl_op_definition
+class NEGOp(Riscv1Rd1RsOperation):
+    name = "riscv.neg"
+
+
+@irdl_op_definition
+class NEGWOp(Riscv1Rd1RsOperation):
+    name = "riscv.negw"
+
+
+@irdl_op_definition
+class SEQZOp(Riscv1Rd1RsOperation):
+    name = "riscv.seqz"
+
+
+@irdl_op_definition
+class SNEZOp(Riscv1Rd1RsOperation):
+    name = "riscv.snez"
+
+
+@irdl_op_definition
+class SLTZOp(Riscv1Rd1RsOperation):
+    name = "riscv.sltz"
+
+
+@irdl_op_definition
+class SGTZOp(Riscv1Rd1RsOperation):
+    name = "riscv.sgtz"
+
+
+@irdl_op_definition
+class BEQZOp(Riscv1Rs1OffOperation):
+    name = "riscv.beqz"
+
+
+@irdl_op_definition
+class BNEZOp(Riscv1Rs1OffOperation):
+    name = "riscv.bnez"
+
+
+@irdl_op_definition
+class BLEZOp(Riscv1Rs1OffOperation):
+    name = "riscv.blez"
+
+
+@irdl_op_definition
+class BGEZOp(Riscv1Rs1OffOperation):
+    name = "riscv.bgez"
+
+
+@irdl_op_definition
+class BLTZOp(Riscv1Rs1OffOperation):
+    name = "riscv.bltz"
+
+
+@irdl_op_definition
+class BGTZOp(Riscv1Rs1OffOperation):
+    name = "riscv.bgtz"
+
+
+@irdl_op_definition
+class BGTOp(Riscv1Rs1Rt1OffOperation):
+    name = "riscv.bgt"
+
+
+@irdl_op_definition
+class BLEOp(Riscv1Rs1Rt1OffOperation):
+    name = "riscv.ble"
+
+
+@irdl_op_definition
+class BGTUOp(Riscv1Rs1Rt1OffOperation):
+    name = "riscv.bgtu"
+
+
+@irdl_op_definition
+class BLEUOp(Riscv1Rs1Rt1OffOperation):
+    name = "riscv.bleu"
+
+
+@irdl_op_definition
+class RETOp(RiscvNoParamsOperation):
+    name = "riscv.ret"
+
+
+# Auxiliary operations
+
+
+@irdl_op_definition
+class LabelOp(Operation):
+    name = "riscv.label"
+    label: OpAttr[LabelAttr]
+
+    @staticmethod
+    def get(label: Union[str, LabelAttr]) -> LabelOp:
+        return LabelOp.build(attributes={"label": label})
+
+
+@irdl_op_definition
+class DirectiveOp(Operation):
```

Comment:
What is a riscv directive? I couldn't find anything in the spec. Maybe we should add a docstring here that explains what this is used for?

Reply:
Ah, it's basically an assembler directive, right? like `put 40 empty bytes here`?

Reply:
Those are meant for generic assembler directive